sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanProcessor.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.trace.SpanId;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link SpanProcessor} implementation for the traceZ zPage.
+ *
+ * <p>Configuration options for {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * can be read from system properties, environment variables, or {@link java.util.Properties}
+ * objects.
+ *
+ * <p>For system properties and {@link java.util.Properties} objects, {@link
+ * io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor} will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code otel.ssp.export.sampled}: sets whether only sampled spans should be exported.
+ * </ul>
+ *
+ * <p>For environment variables, {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code OTEL_SSP_EXPORT_SAMPLED}: sets whether only sampled spans should be exported.
+ * </ul>
+ */
+@ThreadSafe
+public final class TracezSpanProcessor implements SpanProcessor {
+  private final Map<SpanId, ReadableSpan> runningSpanCache;
+  private final Map<SpanId, ReadableSpan> completedSpanCache;
+  private final boolean sampled;
+
+  /**
+   * Constructor for {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}.
+   *
+   * @param sampled report only sampled spans.
+   */
+  public TracezSpanProcessor(boolean sampled) {
+    runningSpanCache = new HashMap<>();
+    completedSpanCache = new HashMap<>();
+    this.sampled = sampled;
+  }
+
+  @Override
+  public void onStart(ReadableSpan span) {
+    synchronized (this) {
```

Comment:
rather than synchronizing everything, what do you think about using a ConcurrentMap implementation?

Reply:
Yeah, that seems to be better here. However, if we impose a max value on the number of spans, then I'll have to decide how to best evict spans, and I might build my own data structure for that.

Reply:
On second thought, I might use an `EvictingQueue.`

Reply:
`caffeine` is pretty small and a great library, would consider it if we end up with an LRU. We want to keep dependencies light but I would definitely want a good cache implementation backing a span processor that hits every span.